components/brave_ads/browser/notification_helper_mac.mm

```diff
@@ -0,0 +1,115 @@
+#include "chrome/common/chrome_features.h"
+#include "base/logging.h"
+
+#include "brave/components/brave_ads/browser/notification_helper_mac.h"
+#include "chrome/browser/fullscreen.h"
+
+namespace brave_ads {
+
+NotificationHelperMac::NotificationHelperMac() = default;
+
+NotificationHelperMac::~NotificationHelperMac() = default;
+
+bool NotificationHelperMac::ShouldShowNotifications() {
+  if (IsFullScreenMode()) {
+    LOG(WARNING) << "Notification not made: Full screen";
+    return false;
+  }
+
+  if (!base::FeatureList::IsEnabled(features::kNativeNotifications)) {
+    LOG(WARNING) << "Native notification feature is disabled so falling back to"
+        " Message Center";
+    return true;
+  }
+
+  if (!IsNotificationsEnabled()) {
+    LOG(INFO) << "Notification not made: Notifications are disabled";
+    return false;
+  }
+
+  return true;
+}
+
+bool NotificationHelperMac::ShowMyFirstAdNotification() {
+  return false;
+}
+
+bool NotificationHelperMac::CanShowBackgroundNotifications() const {
+  return true;
+}
+
+NotificationHelperMac* NotificationHelperMac::GetInstance() {
+  return base::Singleton<NotificationHelperMac>::get();
+}
+
+NotificationHelper* NotificationHelper::GetInstance() {
+  return NotificationHelperMac::GetInstance();
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+bool NotificationHelperMac::IsNotificationsEnabled() const {
+#if !defined(OFFICIAL_BUILD)
+  // Notification center settings returns UNAuthorizationStatusNotDetermined on
+  // non official builds as development builds are not code signed
+  return true;
+#else
+  if (base::mac::IsAtMostOS10_13()) {
+    // macOS 10.13 and older do not support native notifications. Chromium will
+    // fall back to the Chromium Message Center if native notifications are not
+    // supported
+
+    LOG(WARNING) << "Native notifications are not supported on macOS prior"
+        " to macOS 10.14 so falling back to Message Center";
+    return true;
+  }
+
+  if (@available(macOS 10.14, *)) {
+    __block bool is_authorized = false;
+
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+
+    UNUserNotificationCenter *notificationCenter =
+        [UNUserNotificationCenter currentNotificationCenter];
+
+    [notificationCenter getNotificationSettingsWithCompletionHandler:
+        ^(UNNotificationSettings * _Nonnull settings) {
+      switch (settings.authorizationStatus) {
+        case UNAuthorizationStatusDenied: {
+          LOG(WARNING) << "User notifications authorization status denied";
+          is_authorized = false;
+          break;
+        }
+
+        case UNAuthorizationStatusNotDetermined:
+        case UNAuthorizationStatusAuthorized:
+        case UNAuthorizationStatusProvisional: {
+          is_authorized = true;
+          break;
+        }
+      }
+
+      dispatch_semaphore_signal(semaphore);
+    }];
+
+    dispatch_semaphore_wait(semaphore, DISPATCH_TIME_FOREVER);
```

Comment:
I get nervous about 'FOREVER'... are we _that_ committed to blocking this thread forever if OSX FUBAR's?

Reply:
If macOS foobar's we have bigger problems, and ads are running in their own service